metamath.tex

```diff
@@ -14764,6 +14765,149 @@ \chapter{The MIU System}
+
+/* Axiomatic assertion */
+axiom-stmt ::= LABEL '\$a' typecode MATH-SYMBOL* '\$.'
+
+/* Provable assertion */
+p-statement ::= LABEL '\$p' typecode MATH-SYMBOL* '\$=' proof '\$.'
+
+/* A proof. Note that proofs may be interspersed by comments.
+   If '?' is anywhere in a proof then it is an "incomplete" proof. */
+proof ::= uncompressed-proof | compressed-proof
+
+uncompressed-proof ::= (label | '?')+
+
+compressed-proof ::= '(' label* ')' COMPRESSED-PROOF-BLOCK+
+
+/* Certain constants are used specially as "typecodes" in many
+   databases; for convenience we note that here. */
+typecode ::= constant
+
+filename ::= PRINTABLE-SEQUENCE
+
+constant ::= MATH-SYMBOL
+
+variable ::= MATH-SYMBOL
+\end{verbatim}
+
+\needspace{3\baselineskip}
+Here are the rules for lexical processing (tokenization) beyond
+the constant tokens shown above.
+By convention these tokenization rules have names in upper case.
+Every token is read for the longest possible length.
+Whitespace-separated tokens are read sequentially;
+note that the separating whitespace and \texttt{\$(} ... \texttt{\$)}
+comments are skipped.
+
+If a token definition uses another token definition, the whole thing
+is considered a single token.
+A pattern that is only part of a full token has a name beginning
+with an underscore (``\_'').
+An implementation could tokenize many tokens as a
+\texttt{PRINTABLE-SEQUENCE}
+and then check if it meets the more specific rule shown here.
+
+Comments do not nest, and both \texttt{\$(} and \texttt{\$)}
+have to be surrounded
+by at least one whitespace character (\texttt{\_WHITECHAR}).
+Technically comments end without consuming the trailing
+\texttt{\_WHITECHAR}, but the trailing
+\texttt{\_WHITECHAR} gets ignored anyway so we ignore that detail here.
+Metamath language processors
+are not required to support \texttt{\$)} followed
+immediately by a bare end-of-file, because the closing
+comment symbol is supposed to be followed by a
+\texttt{\_WHITECHAR} such as a newline.
+
+\begin{verbatim}
+PRINTABLE-SEQUENCE ::= _PRINTABLE-CHARACTER+
+
+MATH-SYMBOL ::= (_PRINTABLE-CHARACTER - '\$')+
+
+_PRINTABLE-CHARACTER ::= [#x21-#x7e] /* ASCII printable characters */
+
+LABEL ::= ( _LETTER-OR-DIGIT | '.' | '-' | '_' )+
+
+_LETTER-OR-DIGIT ::= [A-Za-z0-9]
+
+COMPRESSED-PROOF-BLOCK ::= ([A-Z] | '?')+
+
+/* Define whitespace between tokens. The -> SKIP means that when
+   whitespace is seen, it is skipped and we simply read again. */
+WHITESPACE ::= (_WHITECHAR+ | _COMMENT) -> SKIP
+
+/* Comments, which have the form \$( ... \$) and do not nest. */
+_COMMENT ::= '\$(' (_WHITECHAR+ (PRINTABLE-SEQUENCE - '\$)')*
+  _WHITECHAR+ '\$)' _WHITECHAR
+
+/* Whitespace character: (' ' | '\t' | '\r' | '\n' | '\f') */
+_WHITECHAR ::= [#x20#x09#x0d#x0a#x0c]
+\end{verbatim}
+
+This EBNF was developed as a collaboration between
+David A. Wheeler\index{Wheeler, David A.},
+Mario Carneiro\index{Carneiro, Mario}, and
+Benoit Jubin\index{Jubin, Benoit}, inspired by a request
```

Comment:
Thanks for your work.
I would remove ", inspired by a request (and a lot of initial work) by Benoit Jubin".  Actually, I would not mind if that paragraph was removed altogether: DAW and MC already appear as overall contributors of the book, and my contribution is in contrast negligible.

Reply:
I'm okay with removing it, though I'm obviously okay with giving you credit too :-).  I'll remove the whole paragraph.

Reply:
Actually, better idea: I'll turn it into a comment.  It won't show in the printed work, but it'll still be documented. Your contribution was not negligible, even if you think overwise :-).